File: doko/cards.go
package doko

import "fmt"

type Suit int

const (
	Diamonds Suit = iota
	Hearts
	Spades
	Clubs
)

func (s Suit) String() string {
	return [...]string{"Diamonds", "Hearts", "Spades", "Clubs"}[s]
}

type Rank int

const (
	Nine Rank = iota
	Jack
	Queen
	King
	Ten
	Ace
)

func (s Rank) String() string {
	return [...]string{"Nine", "Jack", "Queen", "King", "Ten", "Ace"}[s]
}

func (s Rank) Value() int {
	return [...]int{0, 2, 3, 4, 10, 11}[s]
}

type Card struct {
	suit Suit
	rank Rank
}

func (c Card) IsDulle() bool {
	return c.suit == Hearts && c.rank == Ten
}

func (c Card) IsStandardTrump() bool {
	return c.suit == Diamonds || c.rank == Jack || c.rank == Queen || c.IsDulle()
}

func (c Card) String() string {
	return fmt.Sprintf("%s %s", c.suit, c.rank)
}

func (c Card) Value() int {
	return c.rank.Value()
}

func CreateDeck() []Card {
	var ans [48]Card
	pos := 0
	for suit := Diamonds; suit <= Clubs; suit++ {
		for rank := Nine; rank <= Ace; rank++ {
			ans[pos] = Card{suit, rank}
			ans[pos+1] = Card{suit, rank}
			pos += 2
		}
	}
	return ans[:]
}

File: frontend/src/model/cards.ts
import { Card } from './core';
import { Rank, Suit } from 'api/karlchen_pb';

export const Diamond9: Card = { rank: Rank.NINE, suit: Suit.DIAMONDS };
export const Diamond10: Card = { rank: Rank.TEN, suit: Suit.DIAMONDS };
export const DiamondJ: Card = { rank: Rank.JACK, suit: Suit.DIAMONDS };
export const DiamondQ: Card = { rank: Rank.QUEEN, suit: Suit.DIAMONDS };
export const DiamondK: Card = { rank: Rank.KING, suit: Suit.DIAMONDS };
export const DiamondA: Card = { rank: Rank.ACE, suit: Suit.DIAMONDS };

export const Heart9: Card = { rank: Rank.NINE, suit: Suit.HEARTS };
export const Heart10: Card = { rank: Rank.TEN, suit: Suit.HEARTS };
export const HeartJ: Card = { rank: Rank.JACK, suit: Suit.HEARTS };
export const HeartQ: Card = { rank: Rank.QUEEN, suit: Suit.HEARTS };
export const HeartK: Card = { rank: Rank.KING, suit: Suit.HEARTS };
export const HeartA: Card = { rank: Rank.ACE, suit: Suit.HEARTS };

export const Spade9: Card = { rank: Rank.NINE, suit: Suit.SPADES };
export const Spade10: Card = { rank: Rank.TEN, suit: Suit.SPADES };
export const SpadeJ: Card = { rank: Rank.JACK, suit: Suit.SPADES };
export const SpadeQ: Card = { rank: Rank.QUEEN, suit: Suit.SPADES };
export const SpadeK: Card = { rank: Rank.KING, suit: Suit.SPADES };
export const SpadeA: Card = { rank: Rank.ACE, suit: Suit.SPADES };

export const Club9: Card = { rank: Rank.NINE, suit: Suit.CLUBS };
export const Club10: Card = { rank: Rank.TEN, suit: Suit.CLUBS };
export const ClubJ: Card = { rank: Rank.JACK, suit: Suit.CLUBS };
export const ClubQ: Card = { rank: Rank.QUEEN, suit: Suit.CLUBS };
export const ClubK: Card = { rank: Rank.KING, suit: Suit.CLUBS };
export const ClubA: Card = { rank: Rank.ACE, suit: Suit.CLUBS };

export function cardString({ suit, rank }: Card): string {
  return `${suitString(suit)}${rankString(rank)}`;
}

function rankString(rank: Rank): string {
  switch (rank) {
    case Rank.ACE:
      return 'A';
    case Rank.JACK:
      return 'J';
    case Rank.KING:
      return 'K';
    case Rank.QUEEN:
      return 'Q';
    case Rank.NINE:
      return '9';
    case Rank.TEN:
      return '10';
  }
}

function suitString(suit: Suit): string {
  switch (suit) {
    case Suit.DIAMONDS:
      return '♦';
    case Suit.HEARTS:
      return '♥';
    case Suit.SPADES:
      return '♠';
    case Suit.CLUBS:
      return '♣';
  }
}
